File: 2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.cpp
class Solution {
public:
    int countPalindromicSubsequence(string s) {
        vector<int> first(26, -1), last(26, -1);

        // Store first & last occurrence of each character
        for (int i = 0; i < s.size(); i++) {
            int c = s[i] - 'a';
            if (first[c] == -1) first[c] = i;
            last[c] = i;
        }

        int ans = 0;

        // For each character, count unique middle characters
        for (int c = 0; c < 26; c++) {
            if (first[c] != -1 && last[c] > first[c]) {
                unordered_set<char> st;
                for (int i = first[c] + 1; i < last[c]; i++) {
                    st.insert(s[i]);
                }
                ans += st.size();
            }
        }
        return ans;
    }
};

File: 2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.py
class Solution(object):
    def countPalindromicSubsequence(self, s):
        """
        :type s: str
        :rtype: int
        """
        first = {}
        last = {}

        # Record first and last occurrence of each character
        for i, ch in enumerate(s):
            if ch not in first:
                first[ch] = i
            last[ch] = i

        ans = 0

        # For each character, count unique chars between first and last position
        for ch in first:
            if last[ch] > first[ch]:
                mid_chars = set(s[first[ch] + 1 : last[ch]])
                ans += len(mid_chars)

        return ans
